### PR TITLE
Add coupling to scattering

### PR DESCRIPTION
**Pull request by ffoucart**
## Proposed changes

Add bookkeeping for hydro-MC coupling in diffusion approximation

### Upgrade instructions

None

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by ffoucart** [comment]
Needs to be updated after we finish #5920

**Comment by ffoucart** [comment]
@kidder : I did some changes after rebasing here (mainly moving the coupling function to a separate file), so I kept the changes in a separate PR (hence the failing test due to a "rebase" PR). For some reason I had in mind that you had looked at this, but I don't see any comment so... my memory might be wrong. If you don't care about the difference between the new and old version, I can merge this into a single commit.

**Comment by ffoucart** [comment]
Ok, everything has been merged in one commit now and passes the test; should be ready for review.

File: src/Evolution/Particles/MonteCarlo/CouplingTermsForPropagation.hpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
```
Comment (kidder) [inline src/Evolution/Particles/MonteCarlo/CouplingTermsForPropagation.hpp]: - include only `DataStructures/Tensor/TypeAliases.hpp` and `IndexType.hpp`
- forward declare not_null, DataVector, Packet
- move the includes to the cpp file

File: src/Evolution/Particles/MonteCarlo/CouplingTermsForPropagation.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <vector>
```
Comment (kidder) [inline src/Evolution/Particles/MonteCarlo/CouplingTermsForPropagation.hpp]: array not vector

File: src/Evolution/Particles/MonteCarlo/CouplingTermsForPropagation.hpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Evolution/Particles/MonteCarlo/Packet.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace Particles::MonteCarlo {
+
+/// \brief Contribution to the neutrino-matter coupling
+/// terms from evolving a packet by time dt, given
+/// absorption and scattering opacities absorption_opacity,
```
Comment (kidder) [inline src/Evolution/Particles/MonteCarlo/CouplingTermsForPropagation.hpp]: I think you can shorten this by omitting the text after opacities

File: src/Evolution/Particles/MonteCarlo/CouplingTermsForPropagation.hpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+/// scattering_opacity.
+///
+/// \details We consider total momentum exchances, not
```
Comment (kidder) [inline src/Evolution/Particles/MonteCarlo/CouplingTermsForPropagation.hpp]: exchanges?

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash the changes

**Comment by ffoucart** [comment]
I think all the comments are addressed here.

**Review by kidder — APPROVED** [review]

**Pull request merged**